### PR TITLE
sqs: Clear depduplication cache when FifoQueue is cleared

### PR DESCRIPTION
**Pull request by martin-walsh**
Issue: https://github.com/localstack/localstack/issues/8211

### CODE REVIEW

**Comment by martin-walsh** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by martin-walsh** [comment]
Can't get tests to run again...

**Comment by martin-walsh** [comment]
closed & re-opened to get circle-ci to run...

**Comment by martin-walsh** [comment]
Tests failing unrelated. Looks to be flaky as arm64 test passed
```
FAILED tests/integration/apigateway/test_apigateway_integrations.py::test_put_integration_responses
```

File: tests/integration/test_sqs.py [hunk 1]
```diff
@@ -2487,6 +2487,36 @@ def test_purge_queue_deletes_delayed_messages(self, sqs_create_queue, aws_client
...
+        aws_client.sqs.send_message(
+            QueueUrl=queue_url,
+            MessageBody="message-1",
+            MessageGroupId=group_id,
+            MessageDeduplicationId=dedup_id,
+        )
+
+        aws_client.sqs.purge_queue(QueueUrl=queue_url)
+
+        aws_client.sqs.send_message(
```
Comment (thrau) [inline tests/integration/test_sqs.py:2506]: In the original reproduction template of your issue https://github.com/localstack/localstack/issues/8211 there was an additional step here before the purge (a receive message call) - is this necessary here?

**Review by thrau — COMMENTED** [review]
He Martin @martin-walsh, thanks a lot for the contribution. Great find!

The solution also looks good to me. I just have two questions regarding the test to clarify before we merge. Is the test covering what it should (see comment inline)? And also just to make sure: did you run the test against aws by running the test with `TEST_TARGET=AWS_CLOUD`, this is what the `aws_validated` marker indicates.

**Review by baermat — COMMENTED** [review]
Thank you for your contribution! I agree with  the points @thrau raised, apart from that this change looks good to me :+1:

File: tests/integration/test_sqs.py (hunk 1 above)
Comment (martin-walsh) [reply tests/integration/test_sqs.py:2506]: It isn't necessary. The issue reproduction step was just to illustrate that the message was pushed previously & could be retrieved. That is covered by other tests.

**Comment by martin-walsh** [comment]
> He Martin @martin-walsh, thanks a lot for the contribution. Great find!
> 
> The solution also looks good to me. I just have two questions regarding the test to clarify before we merge. Is the test covering what it should (see comment inline)? And also just to make sure: did you run the test against aws by running the test with `TEST_TARGET=AWS_CLOUD`, this is what the `aws_validated` marker indicates.

Re `aws_validated` should I copied an existing test and modified it. Are you asking me to remove this or test against aws directly ?

**Comment by martin-walsh** [comment]
In the interest of time, as we are in opposite time zones, I tested against aws. Here is the result:

```$ aws sqs create-queue --queue-name "martin-walsh-test-queue.fifo" --attributes FifoQueue=true
{
    "QueueUrl": "https://sqs.us-west-2.amazonaws.com/xxxxxxxxxxxx/martin-walsh-test-queue.fifo"
}
$ aws sqs send-message --queue-url https://sqs.us-west-2.amazonaws.com/xxxxxxxxxxxx/martin-walsh-test-queue.fifo --message-body "Testing 123" --message-group-id "ABC" --message-deduplication-id "123"
{
    "MD5OfMessageBody": "41884e32dd65188232ce22cde06a153d",
    "MessageId": "e0ed2d75-a4bd-4fb5-9b22-212604fc9383",
    "SequenceNumber": "18877567867944687616"
}

$ aws sqs receive-message --queue-url https://sqs.us-west-2.amazonaws.com/xxxxxxxxxxxx/martin-walsh-test-queue.fifo
{
    "Messages": [
        {
            "MessageId": "e0ed2d75-a4bd-4fb5-9b22-212604fc9383",
            "ReceiptHandle": "AQEBGIW+t+2suV1N3cYMZWuqszW32RHZh3RiRSJYBETW2wKFFdb6s2++HvXrWdoIF61PR9hRkuJr2WnvpZPjq6uVUL4pa64oKjMSxMCFKsXq/9+FDslxfpUJF1rPGvCyC024uQ5PDXVsrJvltBSVzfDYgKEbg6E86a1+udkWpSo2koN0FkKO16tNIf2e3TXkG6FeIUB2iTmY/Tj3XoIYkrekq1FnWdglfnHPDqpiHpFgX+D1dtjkLQB3UpL5Uf36wJ0EARG7v+pS4f/pKcpG7ujCTH7zJlmw+n0oLwoF8YVE7P8=",
            "MD5OfBody": "41884e32dd65188232ce22cde06a153d",
            "Body": "Testing 123"
        }
    ]
}

$ aws sqs purge-queue --queue-url https://sqs.us-west-2.amazonaws.com/xxxxxxxxxxxx/martin-walsh-test-queue.fifo
$ aws sqs send-message --queue-url https://sqs.us-west-2.amazonaws.com/xxxxxxxxxxxx/martin-walsh-test-queue.fifo --message-body "Testing 123" --message-group-id "ABC" --message-deduplication-id "123"
{
    "MD5OfMessageBody": "41884e32dd65188232ce22cde06a153d",
    "MessageId": "890a376a-1728-4e85-b689-4f86b3d247af",
    "SequenceNumber": "18877567877623023872"
}
$ aws sqs receive-message --queue-url https://sqs.us-west-2.amazonaws.com/xxxxxxxxxxxx/martin-walsh-test-queue.fifo
{
    "Messages": [
        {
            "MessageId": "890a376a-1728-4e85-b689-4f86b3d247af",
            "ReceiptHandle": "AQEB6GnGYZ5ol9fcl/JxBzt+PUKrlajKMv/wJE8XuRwYJHrSvYDpM6L3mLTErFDB/5iD+MpIjJjKYVRnR7XKG+AEzOZht8f7mSG2cihOV/+FNu0tBcqG4AsIbHnyYG/7dxTeCJDs+nUH3lOmuZzFv4YXeMOsuLqJjOyQSldxCJ0aLQQoGzEQnbBgNudwA7LXeFjhCzciGHthPkRoS9Gvvqx+/mztSDl9v6qPGyFIjGhT4r662DYgXt5P5Qbv73B4uuQDGbRD2CLrMwpJzJxcpkW+WjptXiYNAEyyPxjFaZI+Jxk=",
            "MD5OfBody": "41884e32dd65188232ce22cde06a153d",
            "Body": "Testing 123"
        }
    ]
}
```

It behaves as the change does, and satisfies the `aws_validated` marker.

**Comment by martin-walsh** [comment]
Hi @thrau & @baermat,

Did you guys get a chance to take a look again ?

**Comment by martin-walsh** [comment]
Brought in changes from upstream.

**Comment by martin-walsh** [comment]
I've also run the test with `TEST_TARGET=AWS_CLOUD TEST_PATH="tests/integration/test_sqs.py::TestSqsProvider::test_purge_queue_clears_fifo_deduplication_cache" make test` and it passes.

This is blocking upgrade to localstack v2, so please let me know if there's anything that needs to be done to get this over line.

**Review by thrau — APPROVED** [review]
Hey @martin-walsh , thanks for the contribution and thanks for double checking that the tests run on AWS! :+1:

looks good to me, and welcome to the list of contributors :-)

**Comment by thrau** [comment]
change should be in the latest image in a couple of hours

**Pull request merged**